Micro benchmark for run merging

mheinzel opened the pull request:
Output:

<details>

```
benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x2/merge
time                 12.58 ms   (12.50 ms .. 12.66 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 12.52 ms   (12.48 ms .. 12.58 ms)
std dev              128.4 μs   (88.78 μs .. 215.9 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.496e7    (4.496e7 .. 4.496e7)
  y                  -28.758    (-1143.358 .. 1049.724)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x4/merge
time                 14.25 ms   (14.20 ms .. 14.29 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 14.21 ms   (14.18 ms .. 14.25 ms)
std dev              78.37 μs   (58.54 μs .. 113.9 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.499e7    (4.499e7 .. 4.499e7)
  y                  -136.455   (-1323.932 .. 1207.667)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x7/merge
time                 15.77 ms   (15.71 ms .. 15.82 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 15.75 ms   (15.72 ms .. 15.86 ms)
std dev              133.0 μs   (54.35 μs .. 263.1 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.504e7    (4.504e7 .. 4.504e7)
  y                  3.148      (-60.479 .. 63.803)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x13/merge
time                 17.36 ms   (17.31 ms .. 17.41 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 17.41 ms   (17.37 ms .. 17.52 ms)
std dev              152.7 μs   (67.17 μs .. 292.4 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.517e7    (4.517e7 .. 4.517e7)
  y                  -112.569   (-1013.016 .. 788.339)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-delete-x4/merge
time                 13.63 ms   (13.60 ms .. 13.66 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 13.64 ms   (13.61 ms .. 13.69 ms)
std dev              89.54 μs   (38.76 μs .. 178.3 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.754e7    (3.754e7 .. 3.754e7)
  y                  110.844    (-933.824 .. 1054.549)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-blob-x4/merge
time                 350.7 ms   (350.1 ms .. 351.9 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 352.3 ms   (351.6 ms .. 353.4 ms)
std dev              1.112 ms   (58.81 μs .. 1.388 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              1.567e9    (1.567e9 .. 1.567e9)
  y                  -752.000   (-1504.000 .. 6.898e-6)
variance introduced by outliers: 19% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mupsert-collisions-x4/merge
time                 11.36 ms   (11.33 ms .. 11.39 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 11.40 ms   (11.37 ms .. 11.50 ms)
std dev              128.1 μs   (33.62 μs .. 254.4 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.615e7    (3.614e7 .. 3.615e7)
  y                  -3.538     (-903.746 .. 840.478)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mix-x4/merge
time                 15.83 ms   (15.80 ms .. 15.86 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 15.88 ms   (15.84 ms .. 15.95 ms)
std dev              117.6 μs   (27.07 μs .. 224.0 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.248e7    (4.248e7 .. 4.248e7)
  y                  -296.735   (-1376.169 .. 950.580)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mix-collisions-x4/merge
time                 11.88 ms   (11.86 ms .. 11.90 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 11.92 ms   (11.89 ms .. 11.98 ms)
std dev              97.85 μs   (35.90 μs .. 187.2 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.262e7    (3.262e7 .. 3.262e7)
  y                  454.457    (-842.262 .. 1606.995)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-large-keys-x4/merge
time                 27.80 ms   (27.68 ms .. 27.91 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 27.71 ms   (27.65 ms .. 27.76 ms)
std dev              117.7 μs   (88.09 μs .. 162.6 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              1.707e8    (1.707e8 .. 1.707e8)
  y                  87.111     (-1.524e-6 .. 229.760)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-mixed-vals-x4/merge
time                 28.15 ms   (28.00 ms .. 28.35 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 27.59 ms   (27.33 ms .. 27.79 ms)
std dev              504.4 μs   (376.2 μs .. 619.8 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              1.719e8    (1.719e8 .. 1.719e8)
  y                  40.784     (-1400.984 .. 1500.370)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-page-x4/merge
time                 40.84 ms   (40.52 ms .. 41.17 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 41.58 ms   (41.11 ms .. 42.79 ms)
std dev              1.517 ms   (289.7 μs .. 2.678 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.557e8    (2.557e8 .. 2.557e8)
  y                  91.429     (-2.632e-6 .. 240.000)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-page-plus-byte-x4/merge
time                 75.62 ms   (74.94 ms .. 76.21 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 76.29 ms   (76.02 ms .. 76.62 ms)
std dev              481.6 μs   (346.7 μs .. 682.4 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.914e8    (4.914e8 .. 4.914e8)
  y                  -936.533   (-3067.488 .. 3223.921)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-huge-vals-x4/merge
time                 136.3 ms   (136.0 ms .. 136.8 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 137.2 ms   (136.9 ms .. 138.2 ms)
std dev              860.2 μs   (231.7 μs .. 1.325 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              9.708e8    (9.708e8 .. 9.708e8)
  y                  -155.429   (-310.857 .. 8.762e-6)
variance introduced by outliers: 12% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4/merge
time                 21.11 ms   (21.07 ms .. 21.14 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 21.20 ms   (21.16 ms .. 21.28 ms)
std dev              136.8 μs   (82.43 μs .. 226.4 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              7.367e7    (7.367e7 .. 7.367e7)
  y                  -14.400    (-136.699 .. 92.296)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4-uneven/merge
time                 21.13 ms   (21.09 ms .. 21.16 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 21.22 ms   (21.18 ms .. 21.27 ms)
std dev              97.75 μs   (63.69 μs .. 159.2 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              7.378e7    (7.378e7 .. 7.378e7)
  y                  -276.838   (-1252.356 .. 792.753)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4-lastlevel/merge
time                 15.75 ms   (15.68 ms .. 15.82 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 15.84 ms   (15.80 ms .. 15.92 ms)
std dev              140.5 μs   (89.71 μs .. 215.7 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              5.889e7    (5.889e7 .. 5.889e7)
  y                  211.105    (-7.078 .. 412.992)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4+1-min-skewed-lastlevel/merge
time                 16.18 ms   (16.11 ms .. 16.26 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 16.28 ms   (16.24 ms .. 16.34 ms)
std dev              133.8 μs   (81.17 μs .. 197.3 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              5.906e7    (5.906e7 .. 5.906e7)
  y                  174.265    (-1134.176 .. 1484.813)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4+1-max-skewed-lastlevel/merge
time                 15.55 ms   (15.50 ms .. 15.60 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 15.46 ms   (15.39 ms .. 15.50 ms)
std dev              133.4 μs   (89.37 μs .. 194.7 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              5.913e7    (5.913e7 .. 5.913e7)
  y                  154.180    (-1232.263 .. 1605.893)
```

</details>

jorisdral commented on bench/micro/Bench/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -0,0 +1,391 @@
+    name         :: !String
+    -- | Number of key\/operation pairs for each run.
+  , nentries     :: ![Int]
+  , finserts     :: !Int
+  , fblobinserts :: !Int
+  , fdeletes     :: !Int
+  , fmupserts    :: !Int
```
Also document here that they are frequency weights

jorisdral commented on bench/micro/Bench/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -0,0 +1,391 @@
+          first (serialiseKey . Hash.hash64)
+        . uniformR (0, totalentries `div` 2)
+
+    scaled weights =
+      let total = sum weights
+      in [ round (fromIntegral totalentries * w / total :: Double)
+         | w <- weights
+         ]
```
I find it a slightly confusing way of defining how many entries are in each run. Could we hardcode the number instead of scaling them according to `totalEntries`? It would also probably useful if we can change the total number of entries for a single scenario without changing it for all of them. `randomWord64WithCollisions` can then get an `totalEntries` argument instead

mheinzel replied:
Hardcoding numbers for each run gets a bit cumbersome and makes it harder to see how many entries there are in total. So I still think something like `scaled` is useful. But probably it's better to have `nentries = 50_000 `distributed` [1, 1, 1, 1, 16]`?

jorisdral replied:
Yeah, I like that, thanks!

jorisdral commented on bench/micro/Bench/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -0,0 +1,391 @@
+  where
+    totalentries = 50_000 :: Word64
+
+    -- For each run, a significant part of all possible keys are in use, e.g.
+    -- half of them for 4 runs of equal size. This leads to many collisions.
```
Could you elaborate? What does having 4 runs of equal size have to do with it? And by "in use" do you mean ones that are actually in the run?

mheinzel replied:
Yes, that's what I mean. And the number of runs is relevant because they affect the number of entries per run (the 50k entries get distributed across the runs), and therefore the ratio of keys that are present to possible keys.

I will rework the description, especially since I also added an argument to the function as you suggested.

jorisdral commented on bench/micro/Bench/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -0,0 +1,391 @@
+
+    -- For each run, a significant part of all possible keys are in use, e.g.
+    -- half of them for 4 runs of equal size. This leads to many collisions.
+    -- Note that each input run still has the full number of elements due to the
+    -- use of 'sampleWithoutReplacement'.
```
What full number?

jorisdral commented on bench/micro/Bench/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -0,0 +1,391 @@
+    -- half of them for 4 runs of equal size. This leads to many collisions.
+    -- Note that each input run still has the full number of elements due to the
+    -- use of 'sampleWithoutReplacement'.
+    -- Warning: Using this with a run of around half of all entries (or more)
+    -- leads  to issues with generating distinct keys for all its entries!
```
So, that is a run that contains half of `totalentries`?

jorisdral commented on bench/micro/Bench/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -0,0 +1,391 @@
+          (writeBufferEnv conf)
+          writeBufferEnvCleanup
+
+    -- We'll remove the files on every run, so we can re-use the same run number.
+    getPaths :: IO Run.RunFsPaths
+    getPaths = pure outputRunNumber
+
+    -- We need to keep the other runs, but remove the freshly created one.
+    cleanupPaths :: FS.HasFS IO FS.HandleIO -> Run.RunFsPaths -> IO ()
+    cleanupPaths hasFS paths = do
+        traverse_ (FS.removeFile hasFS) (Run.runFsPaths paths)
+        exists <- FS.doesFileExist hasFS (Run.runChecksumsPath paths)
+        when exists $
+          FS.removeFile hasFS (Run.runChecksumsPath paths)
```
Maybe rename these functions so it's clear that they are only supposed to touch the output run

mheinzel replied:
I simplified that part a bit and called the function `removeOutputRunFiles`.

jorisdral commented on bench/micro/Bench/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -0,0 +1,391 @@
+inputRunNumbers :: [Run.RunFsPaths]
+inputRunNumbers = Run.RunFsPaths <$> [1..]
+
+type InputRuns = [Run (FS.Handle FS.HandleIO)]
```
Not used consistently

jorisdral commented on bench/micro/Bench/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -0,0 +1,391 @@
+  , randomValue  = first serialiseValue . uniform @_ @UTxOValue
+  }
+
+writeBufferEnv ::
```
```suggestion
mergeEnv ::
```

jorisdral commented:
> Output:

It's quite a bit of output. I don't have comments about the individual scenarios, but here are some lower/upper bounds for allocation numbers and running time given the scenarios you've created:
* Between `652` and `31340` bytes allocated per merged entry
* Between `4401408` and `142571` merged entries per second

```sh
ghci> 3.262e7 / 50_000 -- word64-mix-collisions-x4
652.4
ghci> 1.567e9 / 50_000 -- word64-blob-x4
31340.0
ghci> (1000 / 11.36) * 50_000 -- word64-mix-collisions-x4
4401408.450704225
ghci> (1000 / 350.7) * 50_000 -- word64-blob-x4
142571.99885942403
```

I don't think performance statistics for a practical workload would tend to these extremes, and specially not the high end of allocation numbers or the low end of merged-entries-per-second. The "worst-case" scenarios that involve things like large values and blobs are a bit artificial, and we can make them as "worst-case" as we want by increasing the size of blobs or values.

jorisdral commented:
And, if we only consider the UTxO scenarios:
```
ghci> map (\x -> round ((1000 / x) * 50_000)) [21.11, 21.13, 15.75, 16.18, 15.55]
[2368546,2366304,3174603,3090235,3215434]
ghci> map (\x -> round (x / 50_000)) [7.367e7, 7.378e7, 5.889e7, 5.906e7, 5.913e7]
[1473,1476,1178,1181,1183]
```

mheinzel commented:
I addressed all feedback and rebased. I also added two more scenarios:
- only deletes in a last level merge (so nothing gets written at all, purely measures the reading and resolution)
- mupserts without collisions, to make sure they behave like inserts in this case (no extra cost if resolution is not needed)

<details>

<summary>New output</summary>

```
benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x2/merge
time                 12.49 ms   (12.47 ms .. 12.52 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 12.51 ms   (12.49 ms .. 12.57 ms)
std dev              79.93 μs   (31.29 μs .. 155.1 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.495e7    (4.495e7 .. 4.495e7)
  y                  211.544    (-2721.340 .. 2685.224)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x4/merge
time                 13.83 ms   (13.77 ms .. 13.90 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 13.82 ms   (13.80 ms .. 13.85 ms)
std dev              55.92 μs   (36.64 μs .. 83.56 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.498e7    (4.498e7 .. 4.498e7)
  y                  99.385     (-881.350 .. 1079.865)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x7/merge
time                 14.98 ms   (14.95 ms .. 15.00 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 15.04 ms   (15.00 ms .. 15.23 ms)
std dev              168.7 μs   (28.36 μs .. 382.8 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.503e7    (4.503e7 .. 4.503e7)
  y                  170.932    (-997.943 .. 1159.102)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x13/merge
time                 16.36 ms   (16.32 ms .. 16.39 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 16.41 ms   (16.37 ms .. 16.56 ms)
std dev              180.4 μs   (32.17 μs .. 356.2 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.516e7    (4.516e7 .. 4.516e7)
  y                  1136.455   (-1968.839 .. 3977.021)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-delete-x4/merge
time                 12.73 ms   (12.71 ms .. 12.75 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 12.77 ms   (12.75 ms .. 12.85 ms)
std dev              94.36 μs   (23.96 μs .. 187.7 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.754e7    (3.754e7 .. 3.754e7)
  y                  1021.939   (-642.307 .. 2463.027)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-blob-x4/merge
time                 351.6 ms   (350.2 ms .. 353.3 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 352.1 ms   (351.7 ms .. 352.4 ms)
std dev              435.9 μs   (210.7 μs .. 595.2 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              1.567e9    (1.567e9 .. 1.567e9)
  y                  -768.000   (-1536.000 .. 5.099e-6)
variance introduced by outliers: 19% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mupsert-x4/merge
time                 13.98 ms   (13.93 ms .. 14.04 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 13.87 ms   (13.83 ms .. 13.90 ms)
std dev              89.46 μs   (68.04 μs .. 123.9 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.498e7    (4.498e7 .. 4.498e7)
  y                  70.036     (-1079.901 .. 1006.289)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mupsert-collisions-x4/merge
time                 10.63 ms   (10.59 ms .. 10.68 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 10.63 ms   (10.59 ms .. 10.74 ms)
std dev              165.6 μs   (44.59 μs .. 332.4 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.614e7    (3.614e7 .. 3.614e7)
  y                  537.330    (-1894.679 .. 2690.810)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mix-x4/merge
time                 14.88 ms   (14.86 ms .. 14.90 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 14.90 ms   (14.87 ms .. 14.99 ms)
std dev              112.1 μs   (28.71 μs .. 235.2 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.248e7    (4.248e7 .. 4.248e7)
  y                  -20.125    (-57.037 .. 5.502e-7)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mix-collisions-x4/merge
time                 11.18 ms   (11.15 ms .. 11.22 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 11.19 ms   (11.17 ms .. 11.24 ms)
std dev              85.70 μs   (40.96 μs .. 157.5 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.262e7    (3.262e7 .. 3.262e7)
  y                  -11.028    (-32.838 .. 3.700e-7)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-delete-x4-lastlevel/merge
time                 5.992 ms   (5.891 ms .. 6.101 ms)
                     0.997 R²   (0.995 R² .. 0.999 R²)
mean                 6.031 ms   (5.976 ms .. 6.104 ms)
std dev              191.8 μs   (153.8 μs .. 241.7 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.263e7    (2.263e7 .. 2.263e7)
  y                  -342.847   (-3271.842 .. 2730.675)
variance introduced by outliers: 12% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-large-keys-x4/merge
time                 28.43 ms   (28.29 ms .. 28.58 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 28.88 ms   (28.69 ms .. 29.25 ms)
std dev              555.0 μs   (326.2 μs .. 794.0 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              1.707e8    (1.707e8 .. 1.707e8)
  y                  -125.412   (-463.269 .. 103.665)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-mixed-vals-x4/merge
time                 27.21 ms   (27.07 ms .. 27.31 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 27.60 ms   (27.45 ms .. 27.91 ms)
std dev              429.9 μs   (227.0 μs .. 764.7 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              1.719e8    (1.719e8 .. 1.719e8)
  y                  -85.333    (-215.526 .. 1.751e-6)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-page-x4/merge
time                 39.70 ms   (39.53 ms .. 39.83 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 40.22 ms   (39.96 ms .. 40.80 ms)
std dev              707.0 μs   (205.5 μs .. 1.204 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.557e8    (2.557e8 .. 2.557e8)
  y                  779.604    (9.704e-7 .. 1375.269)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-page-plus-byte-x4/merge
time                 72.79 ms   (72.39 ms .. 73.26 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 73.83 ms   (73.38 ms .. 74.59 ms)
std dev              1.012 ms   (572.0 μs .. 1.487 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.914e8    (4.914e8 .. 4.914e8)
  y                  -1.904e-6  (-4.714e-6 .. 4.258e-6)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-huge-vals-x4/merge
time                 133.0 ms   (132.3 ms .. 133.8 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 134.1 ms   (133.5 ms .. 135.0 ms)
std dev              1.014 ms   (247.6 μs .. 1.325 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              9.708e8    (9.708e8 .. 9.708e8)
  y                  -91.429    (-180.594 .. 8.138e-6)
variance introduced by outliers: 12% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4/merge
time                 20.45 ms   (20.38 ms .. 20.51 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 20.54 ms   (20.49 ms .. 20.64 ms)
std dev              162.6 μs   (91.79 μs .. 274.1 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              7.367e7    (7.367e7 .. 7.367e7)
  y                  10.514     (-49.670 .. 65.455)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4-uneven/merge
time                 20.48 ms   (20.41 ms .. 20.54 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 20.57 ms   (20.51 ms .. 20.68 ms)
std dev              195.2 μs   (80.83 μs .. 345.1 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              7.377e7    (7.377e7 .. 7.377e7)
  y                  -15.238    (-41.575 .. 9.438e-7)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4-lastlevel/merge
time                 15.42 ms   (15.36 ms .. 15.45 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 15.45 ms   (15.42 ms .. 15.54 ms)
std dev              116.2 μs   (43.63 μs .. 217.7 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              5.888e7    (5.888e7 .. 5.888e7)
  y                  -775.382   (-2011.799 .. 26.495)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4+1-min-skewed-lastlevel/merge
time                 15.75 ms   (15.68 ms .. 15.80 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 15.85 ms   (15.80 ms .. 15.97 ms)
std dev              191.3 μs   (66.85 μs .. 313.1 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              5.905e7    (5.905e7 .. 5.905e7)
  y                  -188.651   (-1493.805 .. 1095.142)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4+1-max-skewed-lastlevel/merge
time                 15.14 ms   (15.05 ms .. 15.22 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 15.07 ms   (14.98 ms .. 15.15 ms)
std dev              221.6 μs   (145.9 μs .. 350.4 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              5.912e7    (5.912e7 .. 5.912e7)
  y                  288.350    (-781.234 .. 1275.283)
```

</details>

I also calculated the allocations per entry, with some notes:

| benchmark                       | alloc | comment
|---------------------------------|-------|---------------------------------------
| word64-insert-x4:               |   899 | average entry size 16 bytes
| word64-delete-x4:               |   750 |
| word64-blob-x4:                 | 31340 | average blob size 4 kB, we allocate 7 times more!
| word64-mupsert-x4:              |   899 | same as insert
| word64-mupsert-collisions-x4:   |   723 |
| word64-mix-x4:                  |   849 |
| word64-mix-collisions-x4:       |   652 |
| word64-delete-x4-lastlevel:     |   452 | not writing any output
| insert-large-keys-x4:           |  3414 | average key size 2k, some write amplification?
| insert-mixed-vals-x4:           |  3438 | average value size 2k, some write amplification?
| insert-page-x4:                 |  5114 | entry size 4k, some overhead
| insert-page-plus-byte-x4:       |  9828 | 2x write amplification
| insert-huge-vals-x4:            | 19416 | average value size 15k, some overhead
| utxo-x4:                        |  1473 | average entry size 34+60/2=64 bytes, why much more than word64?
| utxo-x4-uneven:                 |  1476 |
| utxo-x4-lastlevel:              |  1178 |
| utxo-x4+1-min-skewed-lastlevel: |  1181 |
| utxo-x4+1-max-skewed-lastlevel: |  1183 |

Generally it seems like we allocate much more than we should. Looking at the cases with large-ish keys and values, it seems like both get copied once. But there's also much more going on, with 900 bytes for really small entries. And UTxO entries only have 26 to 78 bytes more than that. It's not clear why we allocate so much more there.

jorisdral commented on bench/micro/Bench/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -0,0 +1,412 @@
+  , randomValue  :: Rnd SerialisedValue
+  , randomBlob   :: Rnd SerialisedBlob
+  , mergeLevel   :: !Merge.Level
+    -- | Needs to be defined when generating mupserts.
+  , mergeMappend :: !(Maybe Mappend)
```
Looks like you use `const` by default when this is `Nothing`. So the comment doesn't seem entirely true

mheinzel replied:
I need to provide some function when creating the `Merge`, since there can always be mupserts in runs. But having `Nothing` here leads to `entryToUpdateNormal` being used in the input generation, which fails on mupserts.
